### PR TITLE
Always restart JNLP agents

### PR DESCRIPTION
**Pull request by witokondoria**
This PR enables reusing an existing agent on jenkins restarts. It will also do the work if the host running the docker daemon gets restarted

### CODE REVIEW

**Comment by KostyaSha** [comment]
Better expose it in config option. I don't think that build will work after slave disconnected or daemon restarted. Have you tried it?

**Comment by witokondoria** [comment]
It has been tested with:

* a restarted daemon (docker 1.12), working if the jenkins master still keeps the container registered as computer
* a restarted Jenkins master, the agent reconnects and goes on with its work

I can't easily mock a disconnection, but the agent should also reconnect

**Comment by KostyaSha** [comment]
What type of job was used? I think it may survive restarts but only for durable-tasks?

**Comment by witokondoria** [comment]
Sorry, forgot that: its a pipeline from a github-branch-source plugin, indeed with durable tasks over there

**Comment by KostyaSha** [comment]
For static (not oneshot) retention strategy it makes sense to do restart. 
Entry point is here com.github.kostyasha.yad.DockerCloud#runContainer
Please add new field in `com.github.kostyasha.yad.commons.DockerCreateContainer` with custom enum (no docker-java reusage), add in UI field to choose it.

And in JNLP launcher add conditional on restart policy available. 

This will allow JNLP detect that container has autorestart so reconnect makes sense.

**Comment by KostyaSha** [comment]
Or simply add in JNLP reconnect checkbox and in container start policy choose. Even better for flexible configuration.

**Comment by witokondoria** [comment]
I have made changes to accomodate the checkbox you mentioned but, unfortunately, during my manual tests I am getting an error:
'java.io.IOException: remote file operation failed: /home/jenkins/workspace/test at hudson.remoting.Channel@1419d10a:docker-180d70632c75: hudson.remoting.ChannelClosedException: channel is already closed'

The tests is as simple as:

1 boot a local dev instance with mvn -f `yet-another-docker-plugin/pom.xml  hpi:run`
2 add the pipeline plugin plugins
3 configure a docker cloud pointing to tcp://localhost:2375, where my local daemon is listening
4 add a new pipeline job with 
`node ("restart") {
    sh "pwd && date && sleep 180 && date"
}`
5 while in the sleep, perform a `systemctl restart docker.service`
6 watch the IOException appear

As alternatives, I have bumped the Jenkins version from 2.19.4 to 2.58, dragging the remoting dependency from 2.62.3 to 3.7, so tests could be performed with different JNLP protocol versions (3 & 4). The same error happens .

The real disturbing point is that when I opened this PR a month ago, the agent could get disconected (forced with a docker service restart) and after it connected back, the job would continue running and no exception arose. This leds me into thinking that another change has been made to the plugin (or dependencies) causing the 'disconnect and terminate' events.

@KostyaSha, do you have any pointer so that I could go on with this PR? 

PS: Sorry for such a long comment

**Comment by KostyaSha** [comment]
> 5 while in the sleep, perform a systemctl restart docker.service

Maybe init script in container is bad?

**Comment by witokondoria** [comment]
Im working with kostyasha/jenkins-slave:jdk-wget. Forgot to mention that the container indeed gets reconnected as a Jenkins agent an no errors are shown at the container log. Will try to trace timestamps for the IOException and connection events. Will also try to apply the current changes to a previous commit,,,

**Comment by KostyaSha** [comment]
btw, alpine jdk should work also now. And i finally found how to do autobuilds https://hub.docker.com/r/kostyasha/yet-another-docker-plugin/~/dockerfile/ but still looks ugly :)

**Comment by witokondoria** [comment]
Misttery solved (partially). This PR will handle Jenkins master restarts. It should also work with agent disconnections, thanks to [workflow-durable-task-step-plugin](https://github.com/jenkinsci/workflow-durable-task-step-plugin) if in use (eg: pipeline based jobs).

Weirdly, the use case is covered on an already existing test - [ShellStepTest](https://github.com/jenkinsci/workflow-durable-task-step-plugin/blob/master/src/test/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepTest.java#L245) - that works, where if the agent is a DockerSlave provided via this plugin, it wont work.

Event the run is supposed to go on when the agent comes back, there are is an issue related to this( https://issues.jenkins-ci.org/browse/JENKINS-36013)

As a summary, I would mark this PR as done and keep on following the IOException and exit 1 at the workflow-durable-task-step-plugin

**Comment by witokondoria** [comment]
Mistery solved: restarting an agent (a a docker daemon) is supposed to be taken care with workflow-durable-task-step, with [a test case covering this scenario](https://github.com/jenkinsci/workflow-durable-task-step-plugin/blob/master/src/test/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepTest.java#L245)

Nevertheless, theres an [issue](https://issues.jenkins-ci.org/browse/JENKINS-36013) related to this usage.

I would accept this PR as completed (allowing a jnlp slave to auto reconnect), leaving the agent IOException for further investigation on the workflow-durable-task-step side. WDYT?

**Comment by witokondoria** [comment]
@KostyaSha, any estimates to review this PR?

**Comment by witokondoria** [comment]
PR rebased and squashed

**Comment by KostyaSha** [comment]
I think now @gamma @pronovic should join with windows change for the same options.
@witokondoria could you rebase after #163 ?

**Comment by witokondoria** [comment]
Sure, will do tomorrow

**Comment by witokondoria** [comment]
rebased and included the RESTART_SLAVE option for windows JNLP agents (so @gamma and @pronovic just should/could validate it)

**Comment by pronovic** [comment]
The changes to `DockerComputerJNLPLauncher.java` look reasonable.  However, you also need to tweak `init.ps1` to include the equivalent changes from `init.sh`.   I think something like this should work:

```
if ($RESTART_SLAVE -ne "true") {
   $RUN_CMD="$RUN_CMD -noReconnect"
}
```

Once you apply that change, @louisn and I can build off your branch and test this in our Windows Docker infrastructure.  However, I will need a quick primer on how to set up a test case -- what you want me to configure and what the expected behavior should be.

**Comment by KostyaSha** [comment]
> However, I will need a quick primer on how to set up a test case -- what you want me to configure and what the expected behavior should be.

Not sure what to run, but i think firewall connection to master for some timeout time and then unblock it. Slave should reconnect to master and end build.

**Comment by witokondoria** [comment]
Changes applied to the `init.ps1` file.

A simple test would be setting up a docker cloud with a JNLP docker agent with a `Docker Cloud Retention Strategy` and a 10 idle timeout value, run any job requesting the configured label (so the agent gets provisioned) and restart the Jenkins service. 

Upon restarting, the agent should reconnect and show again its available executor slots.

**Comment by louisn** [comment]
Tests with the YAD plugin built from the restartAlwaysJNLP branch with a Windows 10 agent were successful.  The cloud statistics from Jenkins show executing without the restart function 1st, then with the restart function in the plugin.  Both cases the job was submitted for build and then Jenkins was stopped for 8 minutes, then restarted.

```
Provisioning attempts

Cloud                   Template                        Name                                    Started Provisioning    Launch                  Operation   Completed
Windows Docker Cloud	windows-java:jre1.8.0_91	Windows Docker Cloud-6c112777637a	2017-06-01 19:49	38 sec	7 min 22 sec	15 sec	    2017-06-01 19:57
Windows Docker Cloud	windows-java:jre1.8.0_91	Windows Docker Cloud-e94e14652bfd	2017-06-01 17:24	2 hr 33 min and counting
```

**Comment by KostyaSha** [comment]
@witokondoria looks like 2 options depend on each other? 
What if docker create will have restart policy option and jnlp launcher will unfluence only on jnlp settings (noreconnect)?

**Comment by witokondoria** [comment]
@KostyaSha, you suggest to leave the restart policy always on, for every connection method (ssh & jnlp)? In that case, it won't be configurable.

**Comment by witokondoria** [comment]
@KostyaSha: restart checkbox is at each connection type level (SSH & JNLP). Just for JNLP, it won't set the noreconnect flag.

Not sure if @louisn is also interested in testing this last change as just applies to ssh connections

**Comment by witokondoria** [comment]
@KostyaSha & @louisn could we move on with this PR?

**Comment by witokondoria** [comment]
Rebased from master branch. @KostyaSha, could you pay a visit to this PR?

File: yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/launcher/DockerComputerSSHLauncher.java [hunk 1]
```diff
@@ -68,6 +81,10 @@ public void appendContainerConfig(DockerSlaveTemplate dockerSlaveTemplate, Creat
 
         createCmd.withPortSpecs(sshPort + "/tcp");
 
+        if (restartSlave) {
+            createCmd.withRestartPolicy(RestartPolicy.alwaysRestart());
```
Comment (KostyaSha) [inline yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/launcher/DockerComputerSSHLauncher.java]: if we imagine that docker create allows to have different restart policies i.e. restart 10 times, then ssh setting with checkbox will be not correct. Logically user should enable restart in docker create settings (just docker setting) and in launcher (slave.jar option). Just trying to keep things separate, also that may happen that user will set restart once, but launcher will be invisibly adding setting... that also may look unexpected.

**Comment by KostyaSha** [comment]
imho create container setting should be separate from launcher setting.
Launchers should have `reconnect` checkbox,
create container `restart policy` dropdown.

**Comment by witokondoria** [comment]
Split configurations, each to where it belong (CreateContainer and JNLPLaucher). The problem is that i have been unable to bind the selected value at the dropdownList at DockerCreateContainer/config.groovy (#93) to the corresponding restartPolicy field.

Given the lack of documentation at jelly from groovy, I'm unable to go on. Could you lend me a hand and shed some light on this?. @amuniz gave a hint with [another class](https://github.com/jenkinsci/jenkins/blob/08def67a18eee51de9f3f99bc2a792fee1c160e0/core/src/main/resources/lib/hudson/listScmBrowsers.jelly#L34), but it is a jelly file :S

**Comment by KostyaSha** [comment]
@witokondoria first of all you shouldn't depend directly on docker-java, so you need create jenkins classes. Then you can even use enum, all possible binding examples available in this plugin, just search for related.

**Comment by witokondoria** [comment]
Wheres the dependency on docker-java? It is on the shaded one (as the rest of the project)
Want to create an enum for the four restart policies? It is true that the list with policies at the config.groovy is not the leanest, but the opposite seems overkill.

**Comment by KostyaSha** [comment]
docker-java may rename and break binary compatibility it any time and then shaded classes will dissappear and break jenkins configuration. That's why i'm not depending directly on library. simple enum should be enough i think i had it somewhere as example...

**Comment by KostyaSha** [comment]
example https://github.com/KostyaSha/github-integration-plugin/blob/f7fc17a4c48a9140e38165efd3eb7d02c9c8b5fe/github-pullrequest-plugin/src/main/resources/com/github/kostyasha/github/integration/generic/repoprovider/GitHubPluginRepoProvider/config.groovy#L21

**Comment by KostyaSha** [comment]
When i want to have some nice ui messages - i do objects in enums, when just list something - only list of enums.

**Comment by KostyaSha** [comment]
@witokondoria cleanup and bit reworked, please review https://github.com/KostyaSha/yet-another-docker-plugin/pull/186

**Comment by KostyaSha** [comment]
released

**Pull request closed without merging**